tls/s2n_x509_validator.c

```diff
@@ -213,95 +214,156 @@ int s2n_x509_validator_set_max_chain_depth(struct s2n_x509_validator *validator,
     return 0;
 }
 
+static S2N_RESULT s2n_verify_host_information_san_entry(struct s2n_connection *conn, GENERAL_NAME *current_name, bool *san_found)
+{
+    RESULT_ENSURE_REF(conn);
+    RESULT_ENSURE_REF(current_name);
+    RESULT_ENSURE_REF(san_found);
+
+    if (current_name->type == GEN_DNS || current_name->type == GEN_URI) {
+        *san_found = true;
+
+        const char *name = (const char *) ASN1_STRING_data(current_name->d.ia5);
+        size_t name_len = (size_t) ASN1_STRING_length(current_name->d.ia5);
+
+        RESULT_ENSURE(conn->verify_host_fn(name, name_len, conn->data_for_verify_host), S2N_ERR_CERT_UNTRUSTED);
+
+        return S2N_RESULT_OK;
+    }
+
+    if (current_name->type == GEN_IPADD) {
+        *san_found = true;
+
+        /* try to validate an IP address if it's in the subject alt name. */
+        const unsigned char *ip_addr = current_name->d.iPAddress->data;
+        size_t ip_addr_len = (size_t) current_name->d.iPAddress->length;
+
+        RESULT_STACK_BLOB(address, INET6_ADDRSTRLEN + 1, INET6_ADDRSTRLEN + 1);
+
+        if (ip_addr_len == 4) {
+            RESULT_GUARD(s2n_inet_ntop(AF_INET, ip_addr, &address));
+        } else if (ip_addr_len == 16) {
+            RESULT_GUARD(s2n_inet_ntop(AF_INET6, ip_addr, &address));
+        } else {
+            /* we aren't able to parse this value so skip it */
+            return S2N_RESULT_ERROR;
+        }
+
+        /* strlen should be safe here since we made sure we were null terminated AND that inet_ntop succeeded */
+        const char *name = (const char *) address.data;
+        size_t name_len = strlen(name);
+
+        RESULT_ENSURE(conn->verify_host_fn(name, name_len, conn->data_for_verify_host), S2N_ERR_CERT_UNTRUSTED);
+
+        return S2N_RESULT_OK;
+    }
+
+    /* we don't understand this entry type so skip it */
+    return S2N_RESULT_ERROR;
+}
+
+static S2N_RESULT s2n_verify_host_information_san(struct s2n_connection *conn, X509 *public_cert, bool *san_found)
+{
+    RESULT_ENSURE_REF(conn);
+    RESULT_ENSURE_REF(public_cert);
+    RESULT_ENSURE_REF(san_found);
+
+    *san_found = false;
+
+    DEFER_CLEANUP(STACK_OF(GENERAL_NAME) *names_list = NULL, GENERAL_NAMES_free_pointer);
+
+    names_list = X509_get_ext_d2i(public_cert, NID_subject_alt_name, NULL, NULL);
+
+    int n = sk_GENERAL_NAME_num(names_list);
+    for (int i = 0; i < n; i++) {
+        GENERAL_NAME *current_name = sk_GENERAL_NAME_value(names_list, i);
+
+        /* if we get any valid entries then return */
+        if (s2n_result_is_ok(s2n_verify_host_information_san_entry(conn, current_name, san_found))) {
+            return S2N_RESULT_OK;
+        }
+    }
+
+    /* if an error was set, then don't override it */
+    if (s2n_errno) {
+        return S2N_RESULT_ERROR;
+    }
```

Comment:
Yeah it's a bit unique. I had the errno checking in there for the case where we encounter SAN types we don't support and not clobbering the previous errno value. But it's possible to get to the end of the list and not have any valid types, at which point we do need to set the errno to something.

Also we want to bail at the end in case the list is empty.

Reply:
Ah, if that's the case, s2n_errno being set doesn't mean that en error occurred. It only has meaning if matched with something you know was an error, like posix errno. So what you've got now is equivalent to what I wrote.

If you want to use S2N_ERR_CERT_UNTRUSTED if the end of the list was reached without a match, you'll need to track that separately.

Reply:
I went a bit of a different direction. Let me know if that's clearer.